finality-aleph/src/testing/clique_network.rs

```diff
@@ -0,0 +1,310 @@
+        service.run(rx).await;
+    });
+    // start connecting with the peers
+    let mut peer_ids = Vec::with_capacity(addr.len());
+    for (id, addrs) in addr.into_iter() {
+        interface.add_connection(id.clone(), addrs);
+        peer_ids.push(id);
+    }
+    // peer main loop
+    // we send random messages to random peers
+    // a message is a number in range 0..n_msg
+    // we also keep a list of messages received at least once
+    // on receiving a message we report the total number of distinct messages received so far
+    // the goal is to receive every message at least once
+    tokio::spawn(async move {
+        let mut received: HashSet<usize> = HashSet::with_capacity(n_msg);
+        let mut send_ticker = tokio::time::interval(Duration::from_millis(5));
+        let mut counter: usize = 0;
+        loop {
+            tokio::select! {
+                _ = send_ticker.tick() => {
+                    counter += 1;
+                    // generate random message
+                    let filler_size = match large_message_interval {
+                        Some(lmi) if counter % lmi == 0 => TWICE_MAX_DATA_SIZE,
+                        _ => 0,
+                    };
+                    let data = match corrupted_message_interval {
+                        Some(cmi) if counter % cmi == 0 => MockData::new_undecodable(thread_rng().gen_range(0..n_msg) as u32, filler_size),
+                        _ => MockData::new(thread_rng().gen_range(0..n_msg) as u32, filler_size),
+                    };
+                    // choose a peer
+                    let peer: MockPublicKey = peer_ids[thread_rng().gen_range(0..peer_ids.len())].clone();
+                    // send
+                    interface.send(data, peer);
+                },
+                data = interface.next() => {
+                    // receive the message
+                    let data: MockData = data.expect("next should not be closed");
+                    // mark the message as received, we do not care about sender's identity
+                    received.insert(data.data() as usize);
+                    // report the number of received messages
+                    report.unbounded_send((our_id.clone(), received.len())).expect("should send");
+                },
+            };
+        }
+    });
+}
+
+/// Takes O(n log n) rounds to finish, where n = n_peers * n_msg.
+async fn scenario(
+    n_peers: usize,
+    n_msg: usize,
+    broken_connection_interval: Option<usize>,
+    large_message_interval: Option<usize>,
+    corrupted_message_interval: Option<usize>,
+    status_report_interval: Duration,
+) {
+    // create spawn_handle, we need to keep the task_manager
+    let task_manager =
+        TaskManager::new(Handle::current(), None).expect("should create TaskManager");
+    let spawn_handle = task_manager.spawn_handle();
+    // create peer identities
+    info!(target: LOG_TARGET, "generating keys...");
```

Comment:
I think that tests can have a little different target for clarity

Reply:
Good point, made a separate `const` for them and also for `tcp` (also makes sense to have that be separate) and made the one in `clique` non-public.